libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -0,0 +1,453 @@
+    id: String,
+    sender: Sender<PmMessage>,
+    recv: Receiver<Notification>,
+}
+
+impl NotificationHandler {
+    pub fn listen(&self) -> Result<Option<Notification>, ConnectionManagerError> {
+        match self.recv.try_recv() {
+            Ok(notification) => Ok(Some(notification)),
+            Err(TryRecvError::Disconnected) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
+            )),
+            Err(TryRecvError::Empty) => Ok(None),
+        }
+    }
+
+    pub fn unsubscribe(&self) -> Result<(), ConnectionManagerError> {
+        let message = PmMessage::UnSubscribe(self.id.clone());
+        match self.sender.try_send(message) {
+            Ok(()) => Ok(()),
+            Err(_) => Err(ConnectionManagerError::SendTimeoutError(
+                "Disconnection request timed out".into(),
+            )),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct ConnectionMetadata {
+    id: usize,
+    endpoint: String,
+    ref_count: u64,
+}
+
+struct ConnectionState {
+    connections: HashMap<String, ConnectionMetadata>,
+    mesh: Mesh,
+    transport: Box<dyn Transport + Send>,
+}
+
+impl ConnectionState {
+    fn new(mesh: Mesh, transport: Box<dyn Transport + Send>) -> Self {
+        Self {
+            mesh,
+            transport,
+            connections: HashMap::new(),
+        }
+    }
+
+    fn add_connection(&mut self, endpoint: &str) -> Result<(), ConnectionManagerError> {
+        if let Some(meta) = self.connections.get_mut(endpoint) {
+            meta.ref_count = meta.ref_count + 1;
+        } else {
+            let connection = self.transport.connect(endpoint).map_err(|err| {
+                ConnectionManagerError::ConnectionCreationError(format!("{:?}", err))
+            })?;
+
+            let id = self.mesh.add(connection).map_err(|err| {
+                ConnectionManagerError::ConnectionCreationError(format!("{:?}", err))
+            })?;
+
+            self.connections.insert(
+                endpoint.to_string(),
+                ConnectionMetadata {
+                    id,
+                    endpoint: endpoint.to_string(),
+                    ref_count: 1,
+                },
+            );
+        };
+
+        Ok(())
+    }
+
+    fn remove_connection(
+        &mut self,
+        endpoint: &str,
+    ) -> Result<Option<ConnectionMetadata>, ConnectionManagerError> {
+        if let Some(meta) = self.connections.get_mut(endpoint) {
+            meta.ref_count = meta.ref_count - 1;
+            if meta.ref_count < 1 {
+                self.mesh.remove(meta.id).map_err(|err| {
+                    ConnectionManagerError::ConnectionRemovalError(format!("{:?}", err))
+                })?;
+            }
+
+            Ok(Some(meta.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn reconnect(&mut self, endpoint: &str) -> Result<(), ConnectionManagerError> {
+        self.remove_connection(endpoint)?;
```

Comment:
Should we fail on all errors here?  I think some of the errors may be due to the connection not existing...

Reply:
Yes, we should fail on errors from `remove_connection`. If the connection does not exist it returns `Ok(None)`.

Reply:
Ah, I see, I think I missed that detail.